Brokerages/Oanda/OandaBrokerage.cs

```diff
@@ -0,0 +1,1052 @@
+        public bool RefreshSession()
+        {
+            var raw = "";
+            bool success;
+            lock (_lockAccessCredentials)
+            {
+                //Create the client for connection:
+                var client = new RestClient("https://www.quantconnect.com/terminal/");
+
+                //Create the GET call:
+                var request = new RestRequest("processOanda", Method.GET);
+                request.AddParameter("uid", UserId.ToString(), ParameterType.GetOrPost);
+                request.AddParameter("accessToken", AccessToken, ParameterType.GetOrPost);
+                request.AddParameter("refreshToken", RefreshToken, ParameterType.GetOrPost);
+
+                //Submit the call:
+                var result = client.Execute(request);
+                raw = result.Content;
+
+                //Decode to token response: update internal access parameters:
+                var newTokens = JsonConvert.DeserializeObject<TokenResponse>(result.Content); 
+                if (newTokens != null && newTokens.Success)
+                {
+                    AccessToken = newTokens.AccessToken;
+                    RefreshToken = newTokens.RefreshToken;
+                    _issuedAt = newTokens.IssuedAt;
+                    _lifeSpan = TimeSpan.FromSeconds(newTokens.ExpiresIn);
+                    Log.Trace("SESSION REFRESHED: Access: " + AccessToken + " Refresh: " + RefreshToken + " Issued At: " + _lifeSpan + " JSON>>"
+                        + result.Content);
+                    OnSessionRefreshed(newTokens);
+                    success = true;
+                } 
+                else
+                {
+                    Log.Error("Oanda.RefreshSession(): Error Refreshing Session: URL: " + client.BuildUri(request) + " Response: " + result.Content);
+                    success = false;
+                }
+            }
+
+            if (!success)
+            {
+                // if we can't refresh our tokens then we must stop the algorithm
+                OnMessage(new BrokerageMessageEvent(BrokerageMessageType.Error, "RefreshSession", "Failed to refresh access token: " + raw));
+            }
+
+            return success;
+        }
+
+        /// <summary>
+        /// Event invocator for the SessionRefreshed event
+        /// </summary>
+        protected virtual void OnSessionRefreshed(TokenResponse e)
+        {
+            var handler = SessionRefreshed;
+            if (handler != null) handler(this, e);
+        }
+
+        /// <summary>
+        /// Connects the client to the broker's remote servers
+        /// </summary>
+        public override void Connect()
+        {
+            if (IsConnected) return;
+            RefreshSession();
+        }
+
+        /// <summary>
+        /// Disconnects the client from the broker's remote servers
+        /// </summary>
+        public override void Disconnect()
+        {
+            throw new NotImplementedException();
```

Comment:
let's not throw, we call disconnect when cleaning up resource/disposing, if he has nothing to do here the // NOP is perfectly acceptable.